src/Symfony/Component/String/Utf8String.php

```diff
@@ -0,0 +1,360 @@
+
+    private static $transliterators = [];
+
+    public function __construct(string $string = '')
+    {
+        if ('' !== $string && !preg_match('//u', $string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        $this->string = $string;
+    }
+
+    public static function fromCodePoints(int ...$codes): self
+    {
+        $string = '';
+        foreach ($codes as $code) {
+            if (0x80 > $code %= 0x200000) {
+                $string .= \chr($code);
+            } elseif (0x800 > $code) {
+                $string .= \chr(0xC0 | $code >> 6).\chr(0x80 | $code & 0x3F);
+            } elseif (0x10000 > $code) {
+                $string .= \chr(0xE0 | $code >> 12).\chr(0x80 | $code >> 6 & 0x3F).\chr(0x80 | $code & 0x3F);
+            } else {
+                $string .= \chr(0xF0 | $code >> 18).\chr(0x80 | $code >> 12 & 0x3F).\chr(0x80 | $code >> 6 & 0x3F).\chr(0x80 | $code & 0x3F);
+            }
+        }
+
+        $str = new static();
+        $str->string = $string;
+
+        return $str;
+    }
+
+    public function append(string ...$suffix): parent
+    {
+        $str = clone $this;
+        $str->string .= 1 >= \count($suffix) ? $suffix[0] ?? '' : implode('', $suffix);
+
+        if (!preg_match('//u', $str->string)) {
+            throw new InvalidArgumentException('Invalid UTF-8 string.');
+        }
+
+        return $str;
+    }
+
+    /**
+     * Generic UTF-8 to ASCII transliteration.
+     *
+     * Install the intl extension for best results.
+     *
+     * @param string[] $rules See "*-Latin" rules from Transliterator::listIDs()
+     */
+    public function ascii(array $rules = []): self
+    {
+        $str = clone $this;
+        $s = $str->string;
+        $str->string = '';
+        $step = 0;
+
+        if (\function_exists('transliterator_transliterate')) {
+            $rules[] = 'Any-Latin';
+            $rules[] = 'Latin-ASCII';
```

Comment:
`Latin-ASCII` is not available in all ICU versions, this requires `lib-icu >= 4.6`. We saw several hosting providers still use older ICU versions (https://github.com/ausi/slug-generator/issues/9).

Can be detected with `in_array('Latin-ASCII', \Transliterator::listIDs())`.

Reply:
Thanks for the hint. I found `Latin-ASCII` is a subset of `NFKD`, which we use anyway, so I've removed the rule altogether, it's not needed. For reference, here is the CLDR definition of rules: